Firestore/core/src/firebase/firestore/local/index_manager.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_INDEX_MANAGER_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_INDEX_MANAGER_H_
+
+#include <string>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/model/resource_path.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * Represents a set of indexes that are used to execute queries efficiently.
+ *
+ * Currently the only index is a [collection id] => [parent path] index, used
+ * to execute Collection Group queries.
+ */
+class IndexManager {
+ public:
+  virtual ~IndexManager() {
+  }
+
+  /**
+   * Creates an index entry mapping the collectionId (last segment of the path)
+   * to the parent path (either the containing document location or the empty
+   * path for root-level collections). Index entries can be retrieved via
+   * getCollectionParents().
+   *
+   * NOTE: Currently we don't remove index entries. If this ends up being an
+   * issue we can devise some sort of GC strategy.
+   */
+  virtual void AddToCollectionParentIndex(
+      const model::ResourcePath &collection_path) = 0;
```

Comment:
Nit: in C++ style guide, pointer and reference should stick to the class name, not the variable name, i.e. `ResourcePath& collection_path`, not `ResourcePath &collection_path`. Unfortunately, clang-format just tries to calculate which style is the most prevalent in a given file and format the rest to fit, so this has to be fixed manually.

Applies throughout (as far as C++ files are concerned).

Reply:
Oh, thanks.  Bummer clang-format doesn't enforce it.